### DISCUSSION TITLE
Oracle CURSOR Expressions

### DISCUSSION THREAD

**Discussion by artelk** [linq2db/linq2db#3115] [Q&A] [answered]
With Oracle you can use cursor expressions as column types (see https://docs.oracle.com/cd/B19306_01/server.102/b14200/expressions005.htm).
```sql
SELECT department_name, CURSOR(SELECT salary, commission_pct 
   FROM employees e
   WHERE e.department_id = d.department_id)
   FROM departments d;
```

Is it possible with Linq2Db to somehow make it use cursor expressions?
```c#
var q = from d in db.Departments
            select new
            {
               d.department_name,
               emlpoyees = db.Emlpoyees.Where(e => e.department_id = d.department_id).ToList()
            };
```
This query produces two requests, it queries for Emlpoyees and then for Departments. Then it probably composes the final result.
Naive approach doesn't work (of course):
```c#
[Expression("CURSOR({0})", ServerSideOnly = true]
static IQueryable<T> Cursor<T>(IQueryable<T> q) {...}

var q = from d in db.Departments
            select new
            {
               d.department_name,
               emlpoyees = Cursor(db.Emlpoyees.Where(e => e.department_id = d.department_id)).ToList()
            };
```
What I want to achieve it to have only one roundtrip and to stream the resulting records (having collection fields like that) from the DB (foreach loop, processing the records one by one and immediately releasing them to GC). With two roundtrips it probably fetches the whole result set of the first request before starting the second one and that can be quite memory consuming.

**Comment by artelk** [comment, marked as answer]
It looks it's impossible to read the result of such requests even with ODP.NET.
I'm thinking about converting the cursor expression to xml and deserializing that on the app side:
```sql
SELECT department_name,
  XmlType(CURSOR(SELECT salary, commission_pct FROM employees e WHERE e.department_id = d.department_id)).getClobVal()
FROM departments d;
```
(this throws an error if the nested sql returns an empty set but it is probably possible to fix that).

Is it possible to create an extension to Linq2Db that takes IQueryable, puts it as sql into the XmlType(CURSOR(<_here_>)).getClobVal() and then deserializes the returning clobs to List of objects?

**Reply by sdanyliv** [reply]
Not sure that we will implement that. Too much efforts for single provider.

**Reply by artelk** [reply]
I meant can I do it on app side? 
Linq2Db allows you to create custom extensions for the things not available out of the box (as in https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/Sql/Sql.cs). 
I've tried to write my own static method that takes the IQueryable and put the Expression attribute. But it hanged in runtime. Maybe I did something wrong (not specified some flag/option). Or it is really currently impossible. Can my custom extension expression take IQueryable parameter so it will generate the sql and insert that to the placeholder?

**Reply by sdanyliv** [reply]
It needs support on DataReader level. And even worse we have to handle this case in EagerLoading functionality and make special assumptions.

**Reply by artelk** [reply]
Thanks for answering. 
I believe the support on DataReader level would be needed to be able to read the 'non-standard' value returned by the Db. 
Let me simplify the task: let's assume the expression produces a simple string
```c#
[Expression("XmlType(CURSOR({0})).getStringVal()", ServerSideOnly = true]
static string SubQueryResultSetAsString<T>(IQueryable<T> q) {...} 
```
So it only requires inserting the sql (of the IQueriable q) into the needed place when the request to Db is being generated. Is it possible?

**Reply by artelk** [reply]
Actually it works! 😄 
I just changed the returned type to string and it injected the q-sql into the placeholder. It inserted "select * ..." instead of explicit list of fields, but it's ok. Also there is a limitation on using outer scope fields: it seems in double nested query you cannot use fields from two levels above in Oracle. But you can find workaround.

You've made a really great tool guys!

**Comment by artelk** [comment]
https://gist.github.com/artelk/1ff5c291beb888dcb1a3854d3c0da093
This is my (tricky) solution to avoid the extra roundtrip.
With the SubQueryResultAsList method the subquery records are serialized to xml on the server side and deserialized to list on the application side.
